pkg/http/api.go

```diff
@@ -1412,63 +1446,67 @@ func (a *api) onDirectMessage(reqCtx *fasthttp.RequestCtx) {
 	// Save headers to internal metadata
 	req.WithFastHTTPHeaders(&reqCtx.Request.Header)
 
-	policy := a.resiliency.EndpointPolicy(reqCtx, targetID, fmt.Sprintf("%s:%s", targetID, invokeMethodName))
+	policy := resiliency.NewRunner[*directMessagingPolicyRes](reqCtx,
+		a.resiliency.EndpointPolicy(targetID, targetID+":"+invokeMethodName),
+	)
 	// Since we don't want to return the actual error, we have to extract several things in order to construct our response.
-	var resp *invokev1.InvokeMethodResponse
-	var body []byte
-	var statusCode int
-	var msg ErrorResponse
-	errorOccurred := false
-	err := policy(func(ctx context.Context) (rErr error) {
-		resp, rErr = a.directMessaging.Invoke(ctx, targetID, req)
-
+	dmpr, err := policy(func(ctx context.Context) (*directMessagingPolicyRes, error) {
+		rResp, rErr := a.directMessaging.Invoke(ctx, targetID, req)
 		if rErr != nil {
 			// Allowlists policies that are applied on the callee side can return a Permission Denied error.
 			// For everything else, treat it as a gRPC transport error
-			errorOccurred = true
-			statusCode = fasthttp.StatusInternalServerError
+			invokeErr := invokeError{
+				statusCode: fasthttp.StatusInternalServerError,
+				msg:        NewErrorResponse("ERR_DIRECT_INVOKE", fmt.Sprintf(messages.ErrDirectInvoke, targetID, rErr)),
+			}
 			if status.Code(rErr) == codes.PermissionDenied {
-				statusCode = invokev1.HTTPStatusFromCode(codes.PermissionDenied)
+				invokeErr.statusCode = invokev1.HTTPStatusFromCode(codes.PermissionDenied)
 			}
-			msg = NewErrorResponse("ERR_DIRECT_INVOKE", fmt.Sprintf(messages.ErrDirectInvoke, targetID, rErr))
-			return rErr
+			return nil, invokeErr
 		}
 
-		errorOccurred = false
-		invokev1.InternalMetadataToHTTPHeader(reqCtx, resp.Headers(), reqCtx.Response.Header.Set)
-		var contentType string
-		contentType, body = resp.RawData()
-		reqCtx.Response.Header.SetContentType(contentType)
+		dmpr := &directMessagingPolicyRes{}
+
+		invokev1.InternalMetadataToHTTPHeader(reqCtx, rResp.Headers(), reqCtx.Response.Header.Set)
```

Comment:
qq Is it ok to share `reqCtx` across go routine calls? 
https://pkg.go.dev/github.com/valyala/fasthttp#RequestCtx
`It is unsafe modifying/reading RequestCtx instance from concurrently running goroutines. The only exception is TimeoutError*, which may be called while other goroutines accessing RequestCtx.`

Reply:
Good catch. I'll change how this works so `reqCtx` is not modified within a callback, as the callback may continue even when the response was sent to the user (e.g. in case of timeouts)